File: code/R/ch08/08.03.R
## 8.3 入力変数の数と訓練誤差・検証誤差の関係

library(tidyverse)
library(caret)
#my_data <- read_csv("wine.csv") # 8.1節で作成したwine.csvを使う．
my_url <- "https://raw.githubusercontent.com/taroyabuki/fromzero/master/data/wine.csv"
my_data <- read_csv(my_url)
my_model1 <- train(form = LPRICE2 ~ ., data = my_data, method = "lm")
y_ <- my_model1 %>% predict(my_data)
RMSE(y_, my_data$LPRICE2) # RMSE（訓練）
#> [1] 0.2586167

n <- nrow(my_data) # データの件数
my_data2 <- my_data %>% mutate(rand1 = runif(n)) # 乱数からなる列の追加

my_model2 <- train(form = LPRICE2 ~ ., data = my_data2, method = "lm")

y2_ <- my_model2 %>% predict(my_data2)
RMSE(y2_, my_data$LPRICE2) # RMSE（訓練）
#> [1] 0.2534026

c(my_model1$result$RMSE,  # RMSE（検証）（rand1なしのモデル）
  my_model22$result$RMSE) # RMSE（検証）（rand1ありのモデル）

#> [1] 0.3680248 0.3842703

File: code/py/ch08/08.03.py
## 8.3 標準化

import pandas as pd
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
my_url = ('https://raw.githubusercontent.com'
          '/taroyabuki/fromzero/master/data/wine.csv')
my_data = pd.read_csv(my_url)
X, y = my_data.drop(columns=['LPRICE2']), my_data['LPRICE2']

# StandardScalerで標準化した結果をデータフレームに戻してから描画する．
pd.DataFrame(StandardScaler().fit_transform(X),
             columns=X.columns).boxplot(showmeans=True)

## 8.3 標準化

my_pipeline = Pipeline([
    ('sc', StandardScaler()),
    ('lr', LinearRegression())])
my_pipeline.fit(X, y)

## 8.3 標準化

# 線形回帰の部分だけを取り出す．
my_lr = my_pipeline.named_steps.lr
my_lr.intercept_
#> -1.4517651851851847

pd.Series(my_lr.coef_,
          index=X.columns)
#> WRAIN      0.147741
#> DEGREES    0.398724
#> HRAIN     -0.276802
#> TIME_SV    0.192979
#> dtype: float64

## 8.3 標準化

my_test = [[500, 17, 120, 2]]
my_model.predict(my_test)
#> array([-1.49884253])

